### PR TITLE
A very fast O(logn) repeat impl

### PR DESCRIPTION
**Pull request by satanson**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improved performance of function `repeat`.

Calling O(logn) intead of O(n) times of memcpy by taking advantage of earlier partial memcpy result, rather than copying the original source string. the number of memcpy decreases drastically, so more data will be copied for every calling,  thus give the memcpy more chance to perform SIMD optimization.  google-benchmark that has be conduct shows that this algorithm speed up 5X, comparing with the O(n) memcpy implementation.

- performance test

CPU: Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz
Clang: 10.0.1
Google benchmark: 1.5.1

```
+=========================+==============+==============+==========+
| case(size/repeat_times) | orignal O(N) | fast O(logN) | speedup  |
+=========================+==============+==============+==========+
| 1/1                     | 3.10 ns      | 2.46 ns      | ↑ 1.26   |
| 1/15                    | 36.0 ns      | 17.7 ns      | ↑ 2.03   |
| 1/100                   | 244 ns       | 26.9 ns      | ↑ 9.07   |
| 1/1000                  | 2224 ns      | 50.6 ns      | ↑ 43.95  |
| 1/10000                 | 22521 ns     | 179 ns       | ↑ 125.82 |
+-------------------------+--------------+--------------+----------+
| 10/1                    | 3.04 ns      | 2.39 ns      | ↑ 1.27   |
| 10/15                   | 34.2 ns      | 18.6 ns      | ↑ 1.84   |
| 10/100                  | 233 ns       | 37.0 ns      | ↑ 6.30   |
| 10/1000                 | 2278 ns      | 165 ns       | ↑ 13.81  |
| 10/10000                | 23087 ns     | 2066 ns      | ↑ 11.17  |
+-------------------------+--------------+--------------+----------+
```

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by satanson** [comment]
@akuzm some test results has been added

File: src/Functions/repeat.cpp [hunk 1]
```diff
@@ -124,12 +124,39 @@ struct RepeatImpl
     }
 
 private:
+    // A very fast repeat implementation, only invoke memcpy for O(log(n)) times.
+    // as the calling times decreases, more data will be copied for each memcpy, thus
+    // SIMD optimization will be more efficient.
     static void process(const UInt8 * src, UInt8 * dst, UInt64 size, UInt64 repeat_time)
     {
-        for (UInt64 i = 0; i < repeat_time; ++i)
+        if (BOOST_UNLIKELY(repeat_time <= 0))
```
Comment (amosbird) [inline src/Functions/repeat.cpp:132]: we have `unlikely` primitive.

**Comment by satanson** [comment]
@amosbird fixed

**Comment by alexey-milovidov** [comment]
This PVS-Studio warning is correct:

https://clickhouse-test-reports.s3.yandex.net/16937/3dd4f2fb1e37f21c8244ede8280655034e642125/pvs_studio_report/pvs-studio-html-report/sources/repeat.cpp_86.html#ln149

(just replace 1 to 1ULL)

**Comment by satanson** [comment]
@alexey-milovidov  fixed

**Comment by satanson** [comment]
@alexey-milovidov 
the pr can be merged?

**Comment by satanson** [comment]
@amosbird  what should i do for 'force tests'?

**Comment by amosbird** [comment]
> @amosbird what should i do for 'force tests'?

Nothing. I'd like to check if there are any performance improvements in perf tests.

**Review by CurtizJ — APPROVED** [review]
Ok.

**Comment by CurtizJ** [comment]
Test failures are not related to changes.

**Pull request merged**